msgpack/_unpacker.pyx

```diff
@@ -110,6 +118,40 @@ cdef inline init_ctx(unpack_context *ctx,
 def default_read_extended_type(typecode, data):
     raise NotImplementedError("Cannot decode extended type with typecode=%d" % typecode)
 
+cdef inline int get_data_from_buffer(object obj,
+                                     Py_buffer *view,
+                                     char **buf,
+                                     Py_ssize_t *buffer_len,
+                                     int *new_protocol) except 0:
+    cdef object contiguous
+    cdef Py_buffer tmp
+    if PyObject_CheckBuffer(obj):
+        new_protocol[0] = 1
+        if PyObject_GetBuffer(obj, view, PyBUF_FULL_RO) == -1:
+            raise
+        if view.itemsize != 1:
+            PyBuffer_Release(view)
+            raise BufferError("cannot unpack from multi-byte object")
+        if PyBuffer_IsContiguous(view, 'A') == 0:
+            contiguous = PyMemoryView_GetContiguous(obj, PyBUF_READ, 'C')
+            PyObject_GetBuffer(contiguous, &tmp, PyBUF_SIMPLE)
+            PyBuffer_Release(view)
+            PyBuffer_FillInfo(view, contiguous, tmp.buf, tmp.len, 1, 0)
+            Py_DECREF(contiguous)
```

Comment:
What this does:
- ``PyMemoryView_GetContiguous`` creates a  ``memoryview`` object to a contiguous copy of the non-contiguous memory
- Get a ``Py_buffer`` for this contiguous memory and transfer the information to the target buffer
- Release resources as necessary